gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/Concat.feature

```diff
@@ -0,0 +1,213 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+@StepClassMap @StepConcat
+Feature: Step - concat()
+
+  @GraphComputerVerificationInjectionNotSupported
+  Scenario: g_injectXa_bX_concat
+    Given the empty graph
+    And the traversal of
+      """
+      g.inject("a", "b").concat()
+      """
+    When iterated to list
+    Then the result should be unordered
+      | result |
+      | a |
+      | b |
+
+  @GraphComputerVerificationInjectionNotSupported
+  Scenario: g_injectXa_bX_concat_XcX
+    Given the empty graph
+    And the traversal of
+      """
+      g.inject("a", "b").concat("c")
+      """
+    When iterated to list
+    Then the result should be unordered
+      | result |
+      | ac |
+      | bc |
+
+  @GraphComputerVerificationInjectionNotSupported
+  Scenario: g_injectXa_bX_concat_Xc_dX
+    Given the empty graph
+    And the traversal of
+      """
+      g.inject("a", "b").concat("c", "d")
+      """
+    When iterated to list
+    Then the result should be unordered
+      | result |
+      | acd |
+      | bcd |
+
+  @GraphComputerVerificationInjectionNotSupported
+  Scenario: g_injectXa_bX_concat_Xinject_c_dX
+    Given the empty graph
+    And the traversal of
+      """
+      g.inject("a", "b").concat(__.inject("c", "d"))
+      """
+    When iterated to list
+    Then the result should be unordered
+      | result |
+      | acda |
```

Comment:
@spmallette what is your opinion on this result?

Reply:
not what i would have expected. i think i would have expected:

```
| result |
| ac |
| bc|
```

[looking at other tests](https://github.com/apache/tinkerpop/pull/2099/files/86238327a61b7246aa182b86d7b65c135513929f#diff-3cc4e477a58382ce059167e0cede86838765da192b3b734705e68c5d1364eaceR121), it appears the behavior is to consume the whole traversal rather than `next()`. what was the thinking there? that's sorta different that what child traversals do in Gremlin. 

i guess that still doesn't explain the added "a" and "b" at the end of each result. i'm trying to understand why it does that.

Reply:
The semantics of `inject()` are sometimes hard to think about and I think that's what makes this example look strange. Consider:

```
gremlin> g.inject("a","b").map(inject("c","d"))
==>a
==>b
gremlin> g.inject("a","b").map(inject("c","d").fold())
==>[a,c,d]
==>[b]
gremlin> g.inject("a","b").project("a").by(inject('c','d').fold())
==>[a:[a,c,d]]
==>[a:[b]]
gremlin> g.inject("a","b").union(inject("c","d"))
==>a
==>c
==>d
==>b
gremlin> g.inject("a","b").inject("c","d")
==>c
==>d
==>a
==>b
```

I can't seem to find context where `inject()` resets, so given that, perhaps the expectation for the `concat()` is maybe:

```
| result |
|  ac  |
|  bd |
```

???

Reply:
My understanding of the spec is that concat appends the strings in the arguments together and then with the incoming traverser, with each traverser mapping to the results inside the arguments. 

I'm using TraversalUtil.applyAll to get the result of the child traversal, and In this case, "a" is appended in front of the "c" and "d" injected. Then for inject, it appends the "a" from the parent to the end. Then the map operation will move to the next incoming traverser "b" then does the same thing:
```
g.inject("a", "b").concat(__.inject("c", "d"))
==>acdb
==>bcdb
```

We thought about getting "ac" and "bd", but I wasn't sure if it'd be a scalarMap, since we'll have to keep track of matching traverser from the child traversal to the next incoming traverser for map.

Reply:
Updated to 
```
| result |
|  ac  |
|  bd |
```
(see latest comment).